src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_SUBSET_ITERATORS_H
+#define SHOGUN_SUBSET_ITERATORS_H
+
+#include <shogun/labels/Labels.h>
+#include <shogun/features/Features.h>
+
+namespace shogun
+{
+	template <
+	    typename T, std::enable_if<
+	                    (std::is_base_of<T, CFeatures>::value) ||
+	                    (std::is_base_of<T, CLabels>::value)>* = nullptr>
+	class SubsetIterator
+	{
+	public:
+		SubsetIterator(T* pointer, index_t idx = 0)
+		{
+			if (pointer->get_subset_stack()->get_last_subset() != nullptr)
+				m_subset = pointer->get_subset_stack()
+				               ->get_last_subset()
+				               ->get_subset_idx();
+			m_idx = idx;
+			// need to keep this variable so that the vector destructor isn't
+			// called
+			m_argsorted_subset = CMath::argsort(m_subset);
```

Comment:
will sorting be called multiple times in one range loop?

Reply:
no, this would only be at the start and end when the iterator is instantiated. still need to clean this up more though